Script cleanup -- user friendly :)

mfekadu opened the pull request:
## What's New?
* commented some of our scripts
* removed dead code
* create a build.py

**should make it easier to on-board new devs**



## Type of change (pick-one)
- [x] Bug fix (_non-breaking change which fixes an issue_)
- [x] New feature (_non-breaking change which adds functionality_)
- [ ] Breaking change (_fix or feature that would cause existing functionality to not work as expected_)
- [ ] This change requires a documentation update

## TODO
- [ ] does it work on Windows? (i think so, but need someone to try it)
- [ ] does it work on linux (i think so, but need someone to try it)


## Screenshots
### When all goes well
<img width="914" alt="image" src="https://user-images.githubusercontent.com/11131250/83054431-415c3a80-a007-11ea-87c0-bebe49f6a5b0.png">

### if there's a problem
<img width="914" alt="image" src="https://user-images.githubusercontent.com/11131250/83054645-9a2bd300-a007-11ea-8fea-618578fff8db.png">


### I hope this script proves to be helpful/useful

<img width="1046" alt="image" src="https://user-images.githubusercontent.com/11131250/83055189-5c7b7a00-a008-11ea-8f26-0c88723d27dc.png">

mfekadu committed:
make simple & colorful script for build

mfekadu committed:
comment scripts

mfekadu committed:
remove dead code

mfekadu committed:
whoops, that could've been a bug

mfekadu committed:
a cleaner 'running...' printout

mfekadu commented:
@snekiam 

I made changes to [Dockerfile](https://github.com/calpoly-csai/api/pull/160/files#diff-3254677a7917c6c01f55212f86c57fbf), would this still work?

mfekadu commented:
A good argument can be made to delete the unused scripts

[deploy.sh](https://github.com/calpoly-csai/api/pull/160/files?file-filters%5B%5D=.py&file-filters%5B%5D=.sh#diff-60254338249f657a0a83f98258a56bfe)

[download_nlp_stuff.sh](https://github.com/calpoly-csai/api/pull/160/files?file-filters%5B%5D=.py&file-filters%5B%5D=.sh#diff-ec49a325fb393339f9c47c3f02d40809)

etc...

snekiam commented:
> 
> 
> @snekiam
> 
> I made changes to [Dockerfile](https://github.com/calpoly-csai/api/pull/160/files#diff-3254677a7917c6c01f55212f86c57fbf), would this still work?

Yup, those all look non-breaking to me.

mfekadu committed:
comments && a --no-overwrite-secrets flag

make the setup file a lot easier to skim through with whitespace && sections

another parser.add_argument in the build.py

parser.add_argument for setup script too

mfekadu commented:
## new script flag

<img width="1010" alt="image" src="https://user-images.githubusercontent.com/11131250/83058920-f2fe6a00-a00d-11ea-9b3f-72da736c2820.png">

<img width="1010" alt="image" src="https://user-images.githubusercontent.com/11131250/83058975-090c2a80-a00e-11ea-8ff3-d9b144c4d8a7.png">

mfekadu committed:
specify certificate files config in Dockerfile only

mfekadu committed:
make sure root route always return SUCCESS for health check

https://render.com/docs/zero-downtime-deploys

mfekadu committed:
oops almost forgot colorama

mfekadu commented:
> Your [Render](https://render.com) PR Server at https://nimbus-pr-160.onrender.com is now **live**!

nice! thanks!

mfekadu committed:
delete dead scripts

mfekadu committed:
make sure deploy just ONLY that. DEPLOY.

mfekadu commented:
By the way, I really like this tool by @renderinc because I feel a ton more confident in this pull request when I can see evidence of the PR running live!

mfekadu committed:
format & include GET THE `en_core_web_sm` SPACY MODEL

mfekadu committed:
fix typo

snekiam commented:
> 
> 
> By the way, I really like this tool by @renderinc because I feel a ton more confident in this pull request when I can see evidence of the PR running live!

Maybe I don't understand it completely - I can't get it to load at all (it seems like another cloud to host things in? Don't we already have gcp?)

mfekadu commented:
> > By the way, I really like this tool by @renderinc because I feel a ton more confident in this pull request when I can see evidence of the PR running live!
> 
> Maybe I don't understand it completely - I can't get it to load at all (it seems like another cloud to host things in? Don't we already have gcp?)

I'm just using it for my own deploys, I don't recommend it for the club necessarily because you're right, the club has GCP. My personal computer is running a ton of other software and definitely could use a break. 



Also, if the links were not loading, it's because my scripts were actually bad. Feel free to disregard its comments until I get things working.

**I won't merge until it works**

mfekadu commented:
Hmm...

```
google.auth.exceptions.DefaultCredentialsError: ('Failed to load service account credentials from /opt/render/project/src/auth.json', ValueError('No key could be detected.'))
```

snekiam commented:
> > > By the way, I really like this tool by @renderinc because I feel a ton more confident in this pull request when I can see evidence of the PR running live!
> > 
> > Maybe I don't understand it completely - I can't get it to load at all (it seems like another cloud to host things in? Don't we already have gcp?)
> 
> I'm just using it for my own deploys, I don't recommend it for the club necessarily because you're right, the club has GCP. My personal computer is running a ton of other software and definitely could use a break. 
> 
> 
> 
> Also, if the links were not loading, it's because my scripts were actually bad. Feel free to disregard its comments until I get things working.
> 
> **I won't merge until it works**

Makes sense - I can imagine you're not the only one!

snekiam commented:
> 
> 
> Hmm...
> 
> ```
> google.auth.exceptions.DefaultCredentialsError: ('Failed to load service account credentials from /opt/render/project/src/auth.json', ValueError('No key could be detected.'))
> ```

I..don't know that I've ever seen that before - (unless it might be a missing settings.yaml/pydrive thing?)

mfekadu commented:
> > Hmm...
> > ```
> > google.auth.exceptions.DefaultCredentialsError: ('Failed to load service account credentials from /opt/render/project/src/auth.json', ValueError('No key could be detected.'))
> > ```
> 
> I..don't know that I've ever seen that before - (unless it might be a missing settings.yaml/pydrive thing?)


**The files are definitely there**
```
$ pwd
/opt/render/project/src

$ ls folder_id.txt auth.json credentials.json config.json 
auth.json  config.json  credentials.json  folder_id.txt

$ ls /opt/render/project/src/auth.json
/opt/render/project/src/auth.json
```

mfekadu commented:
### I think this is a `python 3.7` bug? 🐍  🐛 🐞 

<details><summary>full stacktrace</summary>
<p>

```
May 27 01:27:45 PM  [2020-05-27 20:27:45,907] ERROR in app: Exception on /ask [POST]
May 27 01:27:45 PM  Traceback (most recent call last):
  File "/opt/render/project/src/.venv/lib/python3.7/site-packages/google/auth/_default.py", line 125, in _load_credentials_from_file
    credentials = service_account.Credentials.from_service_account_info(info)
  File "/opt/render/project/src/.venv/lib/python3.7/site-packages/google/oauth2/service_account.py", line 201, in from_service_account_info
    info, require=["client_email", "token_uri"]
  File "/opt/render/project/src/.venv/lib/python3.7/site-packages/google/auth/_service_account_info.py", line 55, in from_dict
    signer = crypt.RSASigner.from_service_account_info(data)
  File "/opt/render/project/src/.venv/lib/python3.7/site-packages/google/auth/crypt/base.py", line 114, in from_service_account_info
    info[_JSON_FILE_PRIVATE_KEY], info.get(_JSON_FILE_PRIVATE_KEY_ID)
  File "/opt/render/project/src/.venv/lib/python3.7/site-packages/google/auth/crypt/_python_rsa.py", line 171, in from_string
    raise ValueError("No key could be detected.")
ValueError: No key could be detected.
May 27 01:27:45 PM  
May 27 01:27:45 PM  The above exception was the direct cause of the following exception:
May 27 01:27:45 PM  
May 27 01:27:45 PM  10.104.30.5 - - [27/May/2020:20:27:45 +0000] "POST /ask HTTP/1.1" 500 290 "-" "PostmanRuntime/7.25.0"
May 27 01:27:48 PM  10.100.3.252 - - [27/May/2020:20:27:48 +0000] "GET / HTTP/1.1" 200 37 "-" "Render/1.0"
May 27 01:27:45 PM  Traceback (most recent call last):
  File "/opt/render/project/src/.venv/lib/python3.7/site-packages/flask/app.py", line 2446, in wsgi_app
    response = self.full_dispatch_request()
  File "/opt/render/project/src/.venv/lib/python3.7/site-packages/flask/app.py", line 1951, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/opt/render/project/src/.venv/lib/python3.7/site-packages/flask_cors/extension.py", line 161, in wrapped_function
    return cors_after_request(app.make_response(f(*args, **kwargs)))
  File "/opt/render/project/src/.venv/lib/python3.7/site-packages/flask/app.py", line 1820, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/opt/render/project/src/.venv/lib/python3.7/site-packages/flask/_compat.py", line 39, in reraise
    raise value
  File "/opt/render/project/src/.venv/lib/python3.7/site-packages/flask/app.py", line 1949, in full_dispatch_request
    rv = self.dispatch_request()
  File "/opt/render/project/src/.venv/lib/python3.7/site-packages/flask/app.py", line 1935, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/opt/render/project/src/flask_api.py", line 118, in handle_question
    response = {"answer": nimbus.answer_question(question)}
  File "/opt/render/project/src/nimbus.py", line 33, in answer_question
    ans_dict = self.predict_question(question)
  File "/opt/render/project/src/nimbus.py", line 65, in predict_question
    nlp_props = self.variable_extractor.extract_variables(question)
  File "/opt/render/project/src/nimbus_nlp/variable_extractor.py", line 72, in extract_variables
    request = self.get_prediction(sent)
  File "/opt/render/project/src/nimbus_nlp/variable_extractor.py", line 49, in get_prediction
    predictor = automl_v1.PredictionServiceClient(client_options=options)
  File "/opt/render/project/src/.venv/lib/python3.7/site-packages/google/cloud/automl_v1/gapic/prediction_service_client.py", line 189, in __init__
    address=api_endpoint, channel=channel, credentials=credentials
  File "/opt/render/project/src/.venv/lib/python3.7/site-packages/google/cloud/automl_v1/gapic/transports/prediction_service_grpc_transport.py", line 67, in __init__
    "grpc.max_receive_message_length": -1,
  File "/opt/render/project/src/.venv/lib/python3.7/site-packages/google/cloud/automl_v1/gapic/transports/prediction_service_grpc_transport.py", line 108, in create_channel
    address, credentials=credentials, scopes=cls._OAUTH_SCOPES, **kwargs
  File "/opt/render/project/src/.venv/lib/python3.7/site-packages/google/api_core/grpc_helpers.py", line 177, in create_channel
    credentials, _ = google.auth.default(scopes=scopes)
  File "/opt/render/project/src/.venv/lib/python3.7/site-packages/google/auth/_default.py", line 308, in default
    credentials, project_id = checker()
  File "/opt/render/project/src/.venv/lib/python3.7/site-packages/google/auth/_default.py", line 166, in _get_explicit_environ_credentials
    os.environ[environment_vars.CREDENTIALS]
  File "/opt/render/project/src/.venv/lib/python3.7/site-packages/google/auth/_default.py", line 129, in _load_credentials_from_file
    six.raise_from(new_exc, caught_exc)
  File "<string>", line 3, in raise_from
google.auth.exceptions.DefaultCredentialsError: ('Failed to load service account credentials from /opt/render/project/src/auth.json', ValueError('No key could be detected.'))
May 27 01:27:53 PM  10.100.3.252 - - [27/May/2020:20:27:53 +0000] "GET / HTTP/1.1" 200 37 "-" "Render/1.0"
May 27 01:27:58 PM  10.100.3.252 - - [27/May/2020:20:27:58 +0000] "GET / HTTP/1.1" 200 37 "-" "Render/1.0"
May 27 01:28:03 PM  10.100.3.252 - - [27/May/2020:20:28:03 +0000] "GET / HTTP/1.1" 200 37 "-" "Render/1.0"
```

</p>
</details>

mfekadu commented:
I am about to sound like a hypocrite because I previously said

> **I won't merge until it works**

However, even though there are some bugs with the Google API I believe the script updates in this PR are still good.

@snekiam @Jason-Ku @chidiewenike @richagadgil can I get quick approval to make progress from here?

mfekadu committed:
gce should choose SSL_PORT without affecting local machine choice of PORT

https://github.com/calpoly-csai/nimbus-config-secrets/commit/17129659e97c247f2fd357ade17f48021876f0f0

this way
the environment varables can be set either

1. locally via the nimbus-config-secrets repo instructions

2. in the cloud

either way the env vars will contain PORT and SSL_PORT

when deploying to GCE, the $SSL_PORT  will be chosen and passed into the docker `--build-arg`

when running `./deploy.sh` alone, the default PORT=8080 will be chosen.

this resolves any permission issues related to opening up port 443

mfekadu commented:
## I updated the secrets to go along with:
## 06074b8 and https://github.com/calpoly-csai/nimbus-config-secrets/commit/17129659e97c247f2fd357ade17f48021876f0f0

```
PORT=8080
SSL_PORT=443
```

<img width="746" alt="image" src="https://user-images.githubusercontent.com/11131250/83078945-990e9c00-a02f-11ea-83ab-06a26c24e712.png">
<img width="729" alt="image" src="https://user-images.githubusercontent.com/11131250/83078952-9e6be680-a02f-11ea-8691-45555a9c3c8a.png">


## TODO: make sure the `gce_deploy.sh` still works
## TODO: make sure the `gce.yaml` still works

mfekadu committed:
yaml naml calpoly-csai/nimbus-config-secrets@ac8fcb6

snekiam commented on flask_api.py:
```diff
@@ -78,15 +78,17 @@ def init_nimbus_db():
-        response_code = 42
         response_json = jsonify({"name": "hello {}".format(str(app))})
-        return response_json, response_code
+        return response_json, SUCCESS
```
Definitely better to use real HTTP response codes rather than 42 :).

mfekadu committed:
Update .gitignore

snekiam commented on gce_deploy.sh:
```diff
@@ -16,7 +20,7 @@ docker build -t gcr.io/$GCE_PROJECT/$GCE_INSTANCE-image:$GITHUB_SHA \
           --build-arg GOOGLE_CLOUD_NLP_MODEL_NAME \
           --build-arg GIT_SSH_CERT \
-          --build-arg PORT .
+          --build-arg PORT="$SSL_PORT" .
```
Should this be differently formatted than the other arguments?

snekiam reviewed (commented):
My concern is that this might break things before the demo tomorrow - especially since its touching docker stuff.

snekiam commented:
Also just noticed that run_tests are not passing.

cameron-toy reviewed (approved):
Awesome addition!

mfekadu committed:
resolve failing test

https://github.com/calpoly-csai/api/pull/160#issuecomment-635024894

➜  api git:(mf-script-cleanup) ✗ python -m pytest
============================================= test session starts =============================================
platform darwin -- Python 3.6.8, pytest-5.3.5, py-1.8.1, pluggy-0.13.1
rootdir: /Users/mfekadu/GitHub/api
plugins: pylint-0.15.0
collected 25 items

tests/test_database_wrapper.py ..................                                                       [ 72%]
tests/test_flask_api.py ......                                                                          [ 96%]
tests/test_validators.py .                                                                              [100%]

============================================= 25 passed in 2.35s ==============================================
➜  api git:(mf-script-cleanup) ✗

mfekadu committed:
make executable

mfekadu committed:
en_core_web_sm
